src/references.md

```diff
@@ -17,161 +17,140 @@ Which obey the following rules:
-that this is an inclusive relationship: `x` is a descendant and ancestor of
-itself.
 
-We can then define references as simply *names* for paths. When you create a
-reference, you're declaring that an ownership path exists to this address
-of memory.
 
-Tragically, plenty of data doesn't reside on the stack, and we must also
-accommodate this. Globals and thread-locals are simple enough to model as
-residing at the bottom of the stack (though we must be careful with mutable
-globals). Data on the heap poses a different problem.
 
-If all Rust had on the heap was data uniquely owned by a pointer on the stack,
-then we could just treat such a pointer as a struct that owns the value on the
-heap. Box, Vec, String, and HashMap, are examples of types which uniquely
-own data on the heap.
+# Aliasing
 
-Unfortunately, data on the heap is not *always* uniquely owned. Rc for instance
-introduces a notion of *shared* ownership. Shared ownership of a value means
-there is no unique path to it. A value with no unique path limits what we can do
-with it.
+Variables and pointers are generally said to *alias* if they refer to the same
+location in memory.
 
-In general, only shared references can be created to non-unique paths. However
-mechanisms which ensure mutual exclusion may establish One True Owner
-temporarily, establishing a unique path to that value (and therefore all
-its children). If this is done, the value may be mutated. In particular, a
-mutable reference can be taken.
+> Note: this is the broadest definition of aliasing possible, and the Rust model
+> will likely have a more specific definition that takes things like liveness
+> and mutation into account.
 
-The most common way to establish such a path is through *interior mutability*,
-in contrast to the *inherited mutability* that everything in Rust normally uses.
-Cell, RefCell, Mutex, and RWLock are all examples of interior mutability types.
-These types provide exclusive access through runtime restrictions.
+So why should we care about aliasing?
 
-An interesting case of this effect is Rc itself: if an Rc has refcount 1,
-then it is safe to mutate or even move its internals. Note however that the
-refcount itself uses interior mutability.
+Consider this simple function:
 
-In order to correctly communicate to the type system that a variable or field of
-a struct can have interior mutability, it must be wrapped in an UnsafeCell. This
-does not in itself make it safe to perform interior mutability operations on
-that value. You still must yourself ensure that mutual exclusion is upheld.
+```rust
+fn compute(input: &u32, output: &mut u32) {
+    if *input > 10 {
+        *output = 1;
+    }
+    if *input > 5 {
+        *output *= 2;
+    }
+}
+```
 
+We would *like* to be able to optimize it to the following function:
 
+```rust
+fn compute(input: &u32, output: &mut u32) {
+    let cached_input = *input; // keep input in a register (probably)
```

Comment:
Maybe remove the probably? I think it's implied here that this is an optimization we want made and this isn't a LIR or ASM representation so you can't show the register being kept.